### PR TITLE
[SPARK-34565][SQL] Collapse Window nodes with Project between them

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Extend the `CollapseWindow` rule to collapse `Window` nodes, that have `Project` between them.

### Why are the changes needed?

The analyzer will turn a `dataset.withColumn("colName", expressionWithWindowFunction)` method call to a `Project - Window - Project` chain in the logical plan. When this method is called multiple times in a row, then the projects can block the `Window` nodes from being collapsed by the current `CollapseWindow` rule.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -914,13 +914,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
-        if ps1 == ps2 && os1 == os2 && w1.references.intersect(w2.windowOutputSet).isEmpty &&
-          we1.nonEmpty && we2.nonEmpty &&
-          // This assumes Window contains the same type of window expressions. This is ensured
-          // by ExtractWindowFunctions.
-          WindowFunctionType.functionType(we1.head) == WindowFunctionType.functionType(we2.head) =>
+    case w1 @ Window(we1, _, _, w2 @ Window(we2, _, _, grandChild))
+        if windowsCompatible(w1, w2) =>
       w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+
+    case w1 @ Window(we1, _, _, Project(pl, w2 @ Window(we2, _, _, grandChild)))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:921]: What happens if `w1` uses an attribute by produced by the project?

Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:921]: Ah, you are correct. I added a test case for that.

**Comment by HyukjinKwon** [comment]
retest this please

**Review by HyukjinKwon — COMMENTED** [review]
Seems fine to me

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -914,13 +914,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
+      if windowsCompatible(w1, w2) && w1.references.subsetOf(grandChild.outputSet) =>
+      Project(
+        pl ++ we1.map(_.toAttribute),
+        w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+      )
+  }
+
+  def windowsCompatible(w1: Window, w2: Window): Boolean = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:929]: nit: private

**Review by maropu — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@cloud-fan and @hvanhovell, do you have any comment on this? If there's no comments in few days, I will merge this in.

**Comment by tanelk** [comment]
@HyukjinKwon, this has been aproved for a while now, any change, that we can merge this?

**Comment by cloud-fan** [comment]
`Aggregate` kinds of embed a project list and it's easier to write rules to deal with it. We probably should do the same for `Window`. We hit similar issues in https://github.com/apache/spark/pull/31691/files#diff-ce15d5a4cc87dff76686e359221425fb5df32b114a9b2aec5ea18b66d60c0148R49 that we have to take extra care of Project + Window.

**Review by wangyum — COMMENTED** [review]
@tanelk Could you verify is this optimize can benefit TPC-DS q47 and TPC-DS q57?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -914,13 +914,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
+        if windowsCompatible(w1, w2) =>
       w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+
+    case w1 @ Window(we1, _, _, Project(pl, w2 @ Window(we2, _, _, grandChild)))
+      if windowsCompatible(w1, w2) && w1.references.subsetOf(grandChild.outputSet) =>
+      Project(
+        pl ++ we1.map(_.toAttribute),
+        w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+      )
+  }
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:927]: cc @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -914,13 +914,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
+    case w1 @ Window(we1, _, _, w2 @ Window(we2, _, _, grandChild))
+        if windowsCompatible(w1, w2) =>
       w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+
+    case w1 @ Window(we1, _, _, Project(pl, w2 @ Window(we2, _, _, grandChild)))
+      if windowsCompatible(w1, w2) && w1.references.subsetOf(grandChild.outputSet) =>
+      Project(
+        pl ++ w1.windowOutputSet,
+        w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+      )
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:926]:
```scala
      Project(
        pl ++ w1.windowOutputSet,
        w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
      )
```
to
```scala
      Project(
        pl ++ w1.windowOutputSet,
        w1.copy(windowExpressions = we2 ++ we1, child = grandChild))
```
?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -914,13 +914,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
+  }
+
+  private def windowsCompatible(w1: Window, w2: Window): Boolean = {
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:929]: Could we move `windowsCompatible` before `apply`. We usually put the private function before the `apply`. See `TransposeWindow`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -913,14 +913,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
+
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
-        if ps1 == ps2 && os1 == os2 && w1.references.intersect(w2.windowOutputSet).isEmpty &&
-          we1.nonEmpty && we2.nonEmpty &&
-          // This assumes Window contains the same type of window expressions. This is ensured
-          // by ExtractWindowFunctions.
-          WindowFunctionType.functionType(we1.head) == WindowFunctionType.functionType(we2.head) =>
+    case w1 @ Window(we1, _, _, w2 @ Window(we2, _, _, grandChild))
+      if windowsCompatible(w1, w2) =>
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:929]: Two more spaces?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -913,14 +913,27 @@ object OptimizeWindowFunctions extends Rule[LogicalPlan] {
...
       w1.copy(windowExpressions = we2 ++ we1, child = grandChild)
+
+    case w1 @ Window(we1, _, _, Project(pl, w2 @ Window(we2, _, _, grandChild)))
+      if windowsCompatible(w1, w2) && w1.references.subsetOf(grandChild.outputSet) =>
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:933]: Two more spaces?

**Review by HyukjinKwon — APPROVED** [review]
@tanelk, sorry I was stuck in some works. This change looks good to me.

**Comment by HyukjinKwon** [comment]
@hvanhovell does it look good to you too?

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
Another similar issue: https://github.com/apache/spark/pull/31980.

**Comment by tanelk** [comment]
@HyukjinKwon, this has few aprovals, but is there anyone else you would like to take a look at it.

**Comment by tanelk** [comment]
This and a very similar PR #31980 have been approved for a while now. @maropu , could you take another look, maybe we can merge this?

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
Looks fine and no one seems to be against this change, so I'll merge this after the last test run passes.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by tanelk** [comment]
retest this please

**Comment by tanelk** [comment]
@maropu 
when you triggered the tests for this and #31980, the jenkins had some issues.
Now both of these PRs succeeded

**Comment by wangyum** [comment]
retest this please.

**Pull request closed without merging**